No-op Impl Polling Trait

Doordashcon opened the pull request:
updates the `Polling` trait implementation in tests.

bkchr commented:
@Doordashcon do you have any reasoning for this?

Doordashcon commented:
just a follow up on this?
https://github.com/polkadot-fellows/runtimes/pull/347#discussion_r1709679890

bkchr commented:
> @bkchr just a follow up on this?
> [polkadot-fellows/runtimes#347 (comment)](https://github.com/polkadot-fellows/runtimes/pull/347#discussion_r1709679890)

The comment is saying that this no-op should be implemented on `()` or some special `NoOp` struct.

Doordashcon commented:
Implemented `Polling` on `()` and updated the tests where we have methods `unimplemented!()`.

muharem commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -126,7 +126,7 @@ benchmarks_instance_pallet! {
 
 	vote {
-		let class = T::Polls::classes().into_iter().next().unwrap();
+		let class = T::Polls::classes().into_iter().next().ok_or(BenchmarkError::Skip)?;
```
what weight `Skip` will result?

Doordashcon replied:
The Secretary Collective requires skipping these extrinsics to generate weights for the ranked_collective pallet instance which uses this NoOp implementation.

bkchr replied:
Yeah I have not checked the actual code, what we need to do here if `classes()` returns nothing, we need to pass some invalid poll index. The function is still trying to access the storage etc.

Doordashcon commented:
Added `Default` bound.

please review @muharem @bkchr

bkchr commented:
@Doordashcon can you please fix it as done in this pr https://github.com/paritytech/polkadot-sdk/pull/3049?

Doordashcon commented:
> @Doordashcon can you please fix it as done in this pr #3049?

updated other benchmarks as well, hopefully ready for review @muharem @bkchr 🙏

Doordashcon commented:
`frame-omni-bencher` results on the Secretary Collective.

```bash
Pallet: "pallet_ranked_collective", Extrinsic: "vote", Lowest values: [], Highest values: [], Steps: 50, Repeat: 20
Raw Storage Info
========
Storage: `SecretaryCollective::Members` (r:1 w:0)
Proof: `SecretaryCollective::Members` (`max_values`: None, `max_size`: Some(42), added: 2517, mode: `MaxEncodedLen`)

....

Pallet: "pallet_ranked_collective", Extrinsic: "cleanup_poll", Lowest values: [], Highest values: [], Steps: 50, Repeat: 20
Raw Storage Info
========
Storage: `SecretaryCollective::VotingCleanup` (r:1 w:0)
Proof: `SecretaryCollective::VotingCleanup` (`max_values`: None, `max_size`: Some(114), added: 2589, mode: `MaxEncodedLen`)
```

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
-	verify {
+
+		// Dispatch the call and measure execution.
+		#[block]
+		{
+			call.dispatch_bypass_filter(origin)?;
+		}
```
Please do this everywhere.

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+		// Convert `r` to a rank and create members.
 		let rank = r as u16;
-		let first = make_member::<T, I>(rank);
+		let _first = make_member::<T, I>(rank);
```
If this is not needed, it should be removed.

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
-		assert_last_event::<T, I>(Event::MemberRemoved { who, rank }.into());
+
+		// Ensure the correct event was emitted for the member removal.
+		assert_last_event::<T, I>(Event::MemberRemoved { who: who.clone(), rank }.into());
```
```suggestion
		assert_last_event::<T, I>(Event::MemberRemoved { who, rank }.into());
```

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
-		assert_last_event::<T, I>(Event::RankChanged { who, rank: rank + 1 }.into());
+
+		// Ensure the correct event was emitted for the rank change.
+		assert_last_event::<T, I>(Event::RankChanged { who: who.clone(), rank: rank + 1 }.into());
```
```suggestion
		assert_last_event::<T, I>(Event::RankChanged { who, rank: rank + 1 }.into());
```

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+		// Ensure the correct event was emitted depending on the member's rank.
+		assert_last_event::<T, I>(
+			match rank {
+				0 => Event::MemberRemoved { who: who.clone(), rank: 0 },
+				r => Event::RankChanged { who: who.clone(), rank: r - 1 },
```
```suggestion
				0 => Event::MemberRemoved { who, rank: 0 },
				r => Event::RankChanged { who, rank: r - 1 },
```

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+		} else {
+			assert_err!(vote_false, crate::Error::<T, I>::NotPolling);
+		}
+
+		// Vote again with a different decision (false).
+		assert_ok!(Pallet::<T, I>::vote(SystemOrigin::Signed(caller.clone()).into(), poll, false));
```
Copy paste mistake?

Doordashcon replied:
oversight, it's the logic before testing with the secretary collective.

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,304 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+
+		// Ensure the correct event was emitted.
+		assert_has_event::<T, I>(
+			Event::MemberExchanged { who: who.clone(), new_who: new_who.clone() }.into(),
```
```suggestion
			Event::MemberExchanged { who, new_who }.into(),
```

Or is the macro forcing you to do this cloning?

bkchr reviewed (commented):
Generally looking good, just some nitpicks

Doordashcon commented on substrate/frame/referenda/src/types.rs:
```diff
@@ -139,7 +139,7 @@ pub struct TrackInfo<Balance, Moment> {
 pub trait TracksInfo<Balance, Moment> {
 	/// The identifier for a track.
-	type Id: Copy + Parameter + Ord + PartialOrd + Send + Sync + 'static + MaxEncodedLen;
+	type Id: Copy + Parameter + Ord + PartialOrd + Send + Sync + 'static + MaxEncodedLen + Default;
```
`Id` -----> tracks, tracks -----> classes in `pallet_referenda`'s implementation of the `Polling` trait.

https://github.com/paritytech/polkadot-sdk/blob/c987da33935898cd5b2f8605d548bc48727c1815/substrate/frame/referenda/src/lib.rs#L750-L752

ggwpez replied:
This is a breaking change (major bump in the prdoc), since it can break existing code.  
If it is something that is only needed for benchmarking, then the benchmark can use a where clause to only work if the Id is Default, but not requiring it on the trait itself. (the trait does not care about benchmarks).

For example like here:  
https://github.com/paritytech/polkadot-sdk/blob/2ab3f03f0bb1773c3799997d541b60e6be5f67bd/bridges/snowbridge/pallets/outbound-queue/src/benchmarking.rs#L19

Doordashcon commented on substrate/frame/support/src/traits/voting.rs:
```diff
@@ -82,9 +82,9 @@ impl<T, P: Polling<T>> sp_runtime::traits::Get<u32> for ClassCountOf<P, T> {
 pub trait Polling<Tally> {
-	type Index: Parameter + Member + Ord + PartialOrd + Copy + HasCompact + MaxEncodedLen;
+	type Index: Parameter + Member + Ord + PartialOrd + Copy + HasCompact + MaxEncodedLen + Default;
 	type Votes: Parameter + Member + Ord + PartialOrd + Copy + HasCompact + MaxEncodedLen;
-	type Class: Parameter + Member + Ord + PartialOrd + MaxEncodedLen;
+	type Class: Parameter + Member + Ord + PartialOrd + MaxEncodedLen + Default;
```
creating invalid class & poll index with the default values of their respective concreate types.

```rust
....
                let (class_exists, class) = T::Polls::classes()
			.into_iter()
			.next()
			.map(|c| (true, c))
			.unwrap_or_else(|| (false, Default::default()));

		// Get rank and create a poll if the class exists.
		let rank = T::MinRankOfClass::convert(class.clone());
		let poll = if class_exists {
			T::Polls::create_ongoing(class.clone())
				.expect("Poll creation should succeed for rank 0")
		} else {
			Default::default()
		};
....
```

`create_ongoing(...)` Errors out in the NoOp implementation &
`classes()` returns no value

Doordashcon commented:
@bkchr @muharem please review

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,263 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+			.unwrap_or_else(|| (false, Default::default()));
 
+		// Convert the class to a rank and create a caller based on that rank.
+		let rank = T::MinRankOfClass::convert(class.clone());
```
You always require that `MinRankOfClass` returns a valid `rank` for the non existing class. Not that convinced about this.

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +58,263 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+			.into_iter()
+			.next()
+			.map(|c| (true, c))
+			.unwrap_or_else(|| (false, Default::default()));
```
Also general if there doesn't exist any class, how can there exist a default value?

Doordashcon replied:
would this also apply to default poll?

Doordashcon replied:
`Default` on `Poll` type should be changed to `From<u32>`

bkchr replied:
Maybe the best is to have some default rank and if the class doesn't exist the default rank is used. (Basically require Default for Rank). Then the rest should be solvable.

Doordashcon replied:
We should consider `max_ongoing` returns the maximum amount of ongoing polls as `u32` in `runtime-benchmarks` feature flag, I think the polling `Index` should bound `From<u32>` by this logic? 

```rust
/// The maximum amount of ongoing polls within any single class. By default it practically
/// unlimited (`u32::max_value()`).
#[cfg(feature = "runtime-benchmarks")]
fn max_ongoing() -> (Self::Class, u32) {
	(Self::classes().into_iter().next().expect("Always one class"), u32::max_value())
}
```

Doordashcon commented:
It is known that both extrinsics each involve a single read operation. This can be manually accounted for, rather than relying on the default value. Would you agree with this approach @bkchr?

bkchr commented:
See [this](https://github.com/paritytech/polkadot-sdk/pull/5311/files/a1d7801c2c2d0938e3da6cad2cbc9caf10a15b7e#r1790874428).

Doordashcon commented:
please review @bkchr

I saw no other way to create an invalid poll for a NoOp Poll than adding `From<u8>` bound to the `Index` type

Doordashcon commented:
Hopefully we get I get to backport the Secretary Collective after this is merged @bkchr @muharem

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +57,277 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
+				Pallet::<T, I>::vote(SystemOrigin::Signed(caller.clone()).into(), poll, true);
+
+			// If the class exists, expect success; otherwise expect a "NotPolling" error.
+			match class {
+				Some(_) => {
+					assert_ok!(vote_result);
+				},
+				None => {
+					assert_err!(vote_result, crate::Error::<T, I>::NotPolling);
+				},
+			}
```
```suggestion
			if class.is_some() {
				assert_ok!(vote_result);
			} else {
				assert_err!(vote_result, crate::Error::<T, I>::NotPolling);
			}
```

bkchr commented on substrate/frame/ranked-collective/src/benchmarking.rs:
```diff
@@ -56,131 +57,277 @@ fn make_member<T: Config<I>, I: 'static>(rank: Rank) -> T::AccountId {
-		let ev = Event::Voted { who: caller, poll, vote: VoteRecord::Nay(1), tally };
-		assert_last_event::<T, I>(ev.into());
+		// Check the result of the negative vote.
+		match class {
+			Some(_) => {
+				assert_ok!(vote_result);
+			},
+			None => {
+				assert_err!(vote_result, crate::Error::<T, I>::NotPolling);
+			},
+		}
```
```suggestion
		if class.is_some() {
			assert_ok!(vote_result);
		} else {
			assert_err!(vote_result, crate::Error::<T, I>::NotPolling);
		}
```

ggwpez commented on substrate/frame/support/src/traits/voting.rs:
```diff
@@ -82,7 +82,7 @@ impl<T, P: Polling<T>> sp_runtime::traits::Get<u32> for ClassCountOf<P, T> {
 
 pub trait Polling<Tally> {
-	type Index: Parameter + Member + Ord + PartialOrd + Copy + HasCompact + MaxEncodedLen;
+	type Index: Parameter + Member + Ord + PartialOrd + Copy + HasCompact + MaxEncodedLen + From<u8>;
```
Also breaking change (major), and maybe also possibly to avoid it ba using the where bound on the benchmark.

bkchr replied:
Good idea with the benchmark where bound.

bkchr replied:
Maybe we don't even need it. Did not spot where this could be required

Doordashcon replied:
> Also breaking change (major), and maybe also possibly to avoid it ba using the where bound on the benchmark.

The where bound could be

``` rust
#[instance_benchmarks(
where <<T as pallet::Config<I>>::Polls as frame_support::traits::Polling<Tally<T, I, pallet::Pallet<T, I>>>>::Index: From<u8>
)]
mod benchmarks {}
```
> Maybe we don't even need it. Did not spot where this could be required

required for 
```rust
<NoOpPoll as Polling<T>>::Index::MAX.into()
```

ggwpez reviewed (approved):
Going to approve since am async, but could still be worth to try and remove the additional added requirements in the traits.

Doordashcon commented:
Thanks for the benchmarking tip @ggwpez should be ready to merge now!

Doordashcon commented:
@bkchr wondering if there is anything more I can do here?
